--output-format=json

samstokes opened the pull request:
This adds a JSON encoding for the messages Bottled Water sends to Kafka, as an alternative to the existing Avro format.  If using this encoding, no schema registry is required.  This is proposed as an implementation for #15.

It also adds a new command-line option `--output-format=[avro|json]` to select the encoding at startup time.

Avro remains the default for backward compatibility, although I think there's an argument to be made (in a separate PR) for changing that, so that the default configuration could run without a schema registry.

### Design

As proposed in #15, this doesn't change the framing protocol between the Postgres extension and the client daemon, which remains Avro.  This means that in JSON mode, the client is no longer simply passing through bytes it received from the extension, but now needs to parse the binary-encoded Avro and convert it to JSON.

This is solved by keeping track of the Avro schemas corresponding to each table.  The client was already keeping track of the topics and registered schema ids for each table, so I refactored that mapping to also remember the actual Avro schemas (as well as the schema ids).  I extracted the mapping from kafka/registry.c into its own component: [kafka/table_mapper.c](https://github.com/samstokes/bottledwater-pg/blob/json/kafka/table_mapper.c).

### Message format

The JSON format is defined by libavro's `avro_value_to_json` function, which produces the [Avro JSON encoding](https://avro.apache.org/docs/1.7.7/spec.html#json_encoding) as defined in the Avro spec.

Examples:

 * `{"id": {"int": 1}}` // an integer key
 * `{"id": {"int": 3}, "title": {"string": "Man Bites Dog"}}` // a row with two fields

I think there's room to improve on the JSON format - the "type annotations" are unwieldy and confusing (though informative), and it contains whitespace, as in the above examples, and so may be rejected by strict JSON parsers.  I would suggest labelling the format as experimental and subject to change, but the same already applies to the Avro format (e.g. #5, #35), and indeed Bottled Water itself already carries an "early software" caveat, so I think that may be redundant.

### Schema registry in JSON mode

I chose to make it an error to specify `--schema-registry` if `--output-format=json`, in order to simplify the configuration space: you either have Avro with a registry, or JSON without.  This assumption is only enforced at startup, not deeply baked into the code (table_mapper.c will operate correctly whether or not a registry is configured).  We might imagine in the future supporting multiple output formats in the same BW instance (maybe with some kind of configuration file mapping tables to topics and formats, with e.g. a `posts_json` topic and a `posts_avro` topic).

ept commented:
Thanks for this great contribution! I like the approach, the table_mapper refactoring is nice, and the command-line integration looks good to me too.

The only substantial query I have is about parsing the Avro binary keys/values into objects (and the schema management associated with doing so) — as pointed out in the comments on individual commits, the protocol client also does this. If I'm reading the code correctly, I think it's unnecessary for table_mapper to duplicate this work; it might be sufficient to pass through the `avro_value_t` objects that have already been parsed.

I'm not a big fan of Avro's JSON format either (the type annotations are an artifact of Avro using unions with null to denote optional fields, which was arguably a design mistake, but it is what it is). But we could fairly easily define our own JSON format, by looking at [how Avro maps `avro_value_t` objects to JSON](https://github.com/apache/avro/blob/trunk/lang/c/src/value-json.c), and doing our own variant thereof that suits us better. Though we could leave the format change to a later PR if you prefer.

ept commented:
Oh, BTW the problem of unnecessary whitespace in the JSON output should be fixed in Avro 1.8.0. Prior versions of Avro bundled an old version of Jansson with the Avro library, but in 1.8.0 we changed it to use the version of libjansson installed on the system (which is typically newer than what was bundled with Avro). And later versions of Jansson seem to no longer include the spurious whitespace by default.

ept commented:
Actually, scrap my last comment. The whitespace has got nothing to do with the Jansson version. Instead, we need to pass the `JSON_COMPACT` option to `json_dumps` ([docs](http://jansson.readthedocs.org/en/2.7/apiref.html#encoding)).

samstokes commented:
Thanks for the review @ept!  I've addressed your minor comments in commits above.

Re JSON: I could have sworn I remembered a time when strict JSON syntax didn't permit whitespace (e.g. considering `{ "name" : "Sam" }` invalid), but now I can't find any source for that ever having been true.  [json.org](http://www.json.org/) explicitly permits whitespace between tokens, as do [RFC 4627](https://tools.ietf.org/html/rfc4627) and [RFC 7159](https://tools.ietf.org/html/rfc7159).  So I've removed the warning from the code: 31997e1  The extra whitespace still makes the encoding less efficient, but JSON is so inefficient anyway (field names, numbers as strings, punctuation etc) that the extra readability is probably worth a few bytes.

Re parsing the Avro schemas and objects: indeed, it is redundant with work already done by the protocol client, as I discovered just after I submitted this PR :)

I do have an [experimental branch](https://github.com/samstokes/bottledwater-pg/tree/json-reuse) (descendant of this branch) that removes the parsing code and just reuses the parsed objects.  However it's exposed a weird bug in the protocol client where in some circumstances it suddenly starts parsing objects incorrectly (e.g. integer values that should be `1` get parsed as `1 - MAXINT`).  I have a symptomatic fix - it seems to be something to do with a double-free of one of the Avro structs - but don't completely understand yet why the bug occurs, so I've parked that branch until I have a chance to figure it out.

Even if the parsing code were reused, it would still be the case that both protocol_client.c and table_mapper.c each maintain a (now identical) mapping from topic to `avro_schema_t`.  I actually think that might be preferable: to do otherwise would require a more intrusive refactoring to expose more of protocol_client's state to the encoding side of things.  It would also be a little awkward to have the topic->schema mapping in one place and the topic->schema_id mapping in another, but moving them together would require moving the schema_id field into protocol_client's mapping, leaving it blank until it was filled in by a registry lookup, further muddying the waters.  So having each component keep its own mapping might not be too bad.  `avro_schema_t` is [just a pointer](https://github.com/apache/avro/blob/7f2afa8/lang/c/src/avro/schema.h#L32) so keeping twice as many isn't a problem, and we can use Avro's refcounting to avoid storing the actual schema twice.

Am I right in thinking that protocol_client.c's parsing of the key and row objects isn't currently used anywhere?  If that's so, then what do you think about merging this branch as-is for the moment, treating the redundant parsing code as "junk DNA"?  I intend to dig in and fix it in a second pass.

samstokes commented:
Since the proposed implementation is known to work, and there seems to be a bug with the existing parsing code (as described above), I think it's worth deferring resolving the redundancy (and dealing with said bug), in order to merge in this useful functionality now.  I've created issue #56 to clean up the redundancy.

ept commented:
Thanks for going ahead and merging this. Yes, we can clear up the redundancy separately in #56.

The only use of the parsed Avro records at the moment is to pretty-print them as JSON in the `bwtest` tool (which simply prints the stream of events to stdout, without involving Kafka). I find that tool quite useful when working on the parts of the code that do not involve Kafka. However, it is a bit silly to always do the parsing, even though the parsed objects are only used on a debugging code path. That probably calls for a refactoring anyway, but I'm not sure what the best structure would be.

The double-freeing issue sounds worrying. If you can manage to track that down, it would be very good to fix that.
